tests/test_lib_tokens_push.py

```diff
@@ -0,0 +1,378 @@
+
+        # The mobile device has not communicated with the backend, yet.
+        # The user is not authenticated!
+        with self.app.test_request_context('/validate/check',
+                                           method='POST',
+                                           data={"user": "cornelius",
+                                                 "realm": self.realm1,
+                                                 "pass": "",
+                                                 "transaction_id": transaction_id}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            jsonresp = json.loads(res.data.decode('utf8'))
+            # Result-Value is false, the user has not answered the challenge, yet
+            self.assertFalse(jsonresp.get("result").get("value"))
+
+        # Now the smartphone communicates with the backend and the challenge in the database table
+        # is marked as answered successfully.
+        challengeobject_list = get_challenges(serial=tokenobj.token.serial,
+                                              transaction_id=transaction_id)
+        challengeobject_list[0].set_otp_status(True)
+
+        with self.app.test_request_context('/validate/check',
+                                           method='POST',
+                                           data={"user": "cornelius",
+                                                 "realm": self.realm1,
+                                                 "pass": "",
+                                                 "state": transaction_id}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            jsonresp = json.loads(res.data.decode('utf8'))
+            # Result-Value is True, since the challenge is marked resolved in the DB
+        self.assertTrue(jsonresp.get("result").get("value"))
+
+    @responses.activate
+    def test_04_api_authenticate_smartphone(self):
+        # Test the /validate/check endpoints and the smartphone endpoint /ttype/push
+        # for authentication
+
+        # get enrolled push token
+        toks = get_tokens(tokentype="push")
+        self.assertEqual(len(toks), 1)
+        tokenobj = toks[0]
+
+        # set PIN
+        tokenobj.set_pin("pushpin")
+        tokenobj.add_user(User("cornelius", self.realm1))
+
+        def check_firebase_params(request):
+            payload = json.loads(request.body)
+            # check the signature in the payload!
+            data = payload.get("message").get("data")
+
+            sign_string = u"{nonce}|{url}|{serial}|{question}|{title}".format(**data)
+            token_obj = get_tokens(serial=data.get("serial"))[0]
+            pem_pubkey = token_obj.get_tokeninfo(PUBLIC_KEY_SERVER)
+            pubkey_obj = load_pem_public_key(str(pem_pubkey), backend=default_backend())
```

Comment:
This should be ``to_bytes`` instead of ``str``. Otherwise, it fails under Python 3 because ``str(pem_pubkey)`` is a (unicode) string, and ``load_pem_public_key`` expects a bytestring :-)

Reply:
done in 9e0e477d